src/EFCore.Relational/Query/SqlExpressions/ValuesExpression.cs

```diff
@@ -57,16 +69,32 @@ public class ValuesExpression : TableExpressionBase
     [EntityFrameworkInternal]
     public ValuesExpression(
         string? alias,
-        IReadOnlyList<RowValueExpression> rowValues,
+        IReadOnlyList<RowValueExpression>? rowValues,
+        SqlParameterExpression? valuesParameter,
         IReadOnlyList<string> columnNames,
-        IReadOnlyDictionary<string, IAnnotation>? annotations)
+        IReadOnlyDictionary<string, IAnnotation>? annotations = null)
         : base(alias, annotations)
     {
-        Check.DebugAssert(
-            rowValues.All(rv => rv.Values.Count == columnNames.Count),
-            "All row values must have a value count matching the number of column names");
+        if (rowValues is not null)
+        {
+            Check.DebugAssert(
+                rowValues.All(rv => rv.Values.Count == columnNames.Count),
+                "All row values must have a value count matching the number of column names");
+        }
+        if (valuesParameter is not null)
```

Comment:
nit: we generally leave an empty line after closing curlies (e.g. here and below)

```suggestion

        if (valuesParameter is not null)
```